tidy3d/components/eme/simulation.py

```diff
@@ -0,0 +1,859 @@
+
+    def _monitor_num_eme_cells(self, monitor: EMEMonitor) -> int:
+        """Total number of EME cells included in monitor based on simulation grid."""
+        return len(self._monitor_eme_cell_indices(monitor=monitor))
+
+    def _monitor_freqs(self, monitor: Monitor) -> List[pd.NonNegativeFloat]:
+        """Monitor frequencies."""
+        if monitor.freqs is None:
+            return self.freqs
+        return np.intersect1d(self.freqs, monitor.freqs)
+
+    def _monitor_num_freqs(self, monitor: Monitor) -> int:
+        """Total number of freqs included in monitor."""
+        return len(self._monitor_freqs(monitor=monitor))
+
+    def _monitor_num_modes(self, monitor: Monitor) -> int:
+        """Total number of modes included in monitor."""
+        sim_max_num_modes = (
+            self.max_port_modes if isinstance(monitor, EMEFieldMonitor) else self.max_num_modes
+        )
+        if not hasattr(monitor, "num_modes") or monitor.num_modes is None:
+            return sim_max_num_modes
+        return min(monitor.num_modes, sim_max_num_modes)
+
+    def _monitor_num_modes_cell(self, monitor: Monitor, cell_index: int) -> int:
+        """Number of modes included in monitor at certain cell_index."""
+        return min(
+            self.eme_grid.mode_specs[cell_index].num_modes, self._monitor_num_modes(monitor=monitor)
+        )
+
+    @cached_property
+    def max_num_modes(self) -> int:
+        """Max number of modes in the simulation."""
+        return np.max([mode_spec.num_modes for mode_spec in self.eme_grid.mode_specs])
+
+    @cached_property
+    def max_port_modes(self) -> int:
+        """Max number of modes at the two ports."""
+        return max(self.eme_grid.mode_specs[0].num_modes, self.eme_grid.mode_specs[-1].num_modes)
+
+    @cached_property
+    def grid(self) -> Grid:
+        """Grid spatial locations and information as defined by `grid_spec`.
+        This is the grid used in the tangential directions
+        as well as the grid used for field monitors.
+        This is distinct from 'eme_grid', which is the grid
+        used for mode solving and EME propagation.
+
+        Returns
+        -------
+        :class:`.Grid`
+            :class:`.Grid` storing the spatial locations relevant to the simulation.
+        """
+
+        # Add a simulation Box as the first structure
+        structures = [Structure(geometry=self.geometry, medium=self.medium)]
+        structures += self.structures
+
+        # make source for autogrid if needed
+        freqs = self.freqs
+        grid_spec = self.grid_spec
+        sources = []
+        if grid_spec.auto_grid_used and grid_spec.wavelength is None:
+            if not np.all(np.isclose(freqs, freqs[0])):
+                raise SetupError(
+                    "Multiple 'sim.freqs' are supplied. Please supply "
+                    "a 'wavelength' value for 'grid_spec' to control automatic "
+                    "grid generation."
+                )
```

Comment:
Can you define an EMESimulation and submit it without ever invoking `.grid`? I'm just thinking whether it could happen that the user provides a faulty grid spec setup, but can still send it to server for solving, where it will just error. If that can happen, then it would probably makes sense to implement this grid check as a validator.

Reply:
added to `sim._post_init_validators`, thanks